2019-08-03-q-and-as-about-nix-for-haskellers.md

```diff
@@ -0,0 +1,692 @@
+
+2.  _Use `cabal` + Nix to build Haskell projects._
+
+    This method uses a combination of `cabal` and Nix.
+
+    In this method, Nix is used to build a GHC package database of all your
+    Haskell dependencies (as well as their required system packages).
+    For instance, if you're developing a Haskell package that uses
+    [`postgresql-simple`](http://hackage.haskell.org/package/postgresql-simple),
+    Nix will compile `postgresql-simple` as well as its Haskell dependencies
+    like `aeson`, `attoparsec`, etc.  It will then produce a GHC package
+    database with all these dependencies available.
+
+    Normally, you will run `nix-shell` to get into a Bash shell with this GHC
+    package database available.  The shell will have the required system
+    libraries available, like `libpq` and `zlib`.  The shell will also provide
+    build tools like `cabal` and `ghcid`.
+
+    From here, you can run `cabal` commands like normal.  `cabal` will be able
+    to see the GHC package database provided by Nix. `cabal` won't have to compile
+    any dependencies when you run commands like `cabal new-build all` for the
+    first time.  The only thing it will ever need to build is your own project.
+
+    This method is nice for interactive, everyday development.
+
+3.  _Use `stack` + Nix to build Haskell projects._
+
+    `stack` has built-in support for working with Nix, although it works
+    slightly different than `cabal`.
+
+    When using `stack` with Nix, you need to do two things.  One is to pass the
+    `--nix` flag to `stack`.  The other is to edit the `stack.yaml` file and
+    specify a location of a Nix file containing a derivation.  This derivation
+    should contain all the system libraries that are dependencies for the
+    Haskell libraries you are using.  For instance, if you are developing a
+    package that depends on the `postgresql-simple` package, the derivation
+    needs to contain the `libpq` and `zlib` system libraries.
+
+    When you run `stack --nix build`, `stack` internally executes `nix-shell`
+    with the derivation you have provided.  It then re-executes itself in this
+    new environment.  Now that it is running in an environment with all the
+    required system libraries, it starts to compile all your Haskell
+    dependencies.  Finally, it compiles your actual project.
+
+    Unlike with `cabal`, `stack` has no way of directly using a GHC package
+    database provided by the user.  Unfortunately, there is no way of building
+    Haskell packages with Nix and making `stack` use them.
+
+    This method is nice for people and teams already using `stack`, but have
+    complicated system dependencies.
+
+    The `stack` documentation has a
+    [nice section](https://docs.haskellstack.org/en/stable/nix_integration/)
+    about integration with Nix.
+
+If you're looking to use Nix for Haskell development, I recommend setting up
+your CI to use `nix-build`, and then using `cabal` + Nix for everyday
```

Comment:
I think here it would be nice to provide link to some config. Ideally Travis CI (it's the most popular). Also, you give suggestion to use nix on CI, but it's not obvious why would want to use `nix` on CI. So far blog post talks about advantages of using nix for development, but CI is a bit different (in my understanding). Or maybe not and there are indeed good reasons to build with nix on CI!